File: Java/ExtensibleCommands/src/main/java/org/extensiblecommands/SequentialCommand.java
package org.extensiblecommands;

/**
 * Implements sequential execution of multiple commands. If any of the sub-commands fails,
 * the command fails.
 */
public class SequentialCommand extends CompositeCommand {
    public SequentialCommand() {
        this("Sequential");
    }

    /**
     * Constructor
     * @param name      Command name
     */
    public SequentialCommand(String name) {
        super(name);
    }

    @Override
    protected void execute() throws Exception {
        for (var subCommand : subCommands) {
            if (getState() == State.Aborted)
                break;

            subCommand.run();

            processAbortAndPauseEvents();

            if (subCommand.getState() == State.Failed || subCommand.getState() == State.Aborted)
                break;
        }
    }
}

File: CSharp/ExtensibleCommands/ExtensibleCommands/SequentialCommand.cs
using System.Collections.Generic;

namespace ExtensibleCommands
{
    /// <summary>
    /// Implements sequential execution of multiple commands. If any of the sub-commands fails, 
    /// the command fails.
    /// </summary>
    public class SequentialCommand : CompositeCommand
    {
        /// <summary> Constructor </summary>
        public SequentialCommand() : this("Sequential") { }

        /// <summary>
        /// Constructor
        /// </summary>
        /// <param name="name">Command name</param>
        public SequentialCommand(string name) : base(name) { }

        protected override void Execute()
        {
            foreach (ICommand subCommand in _subCommands)
            {
                if (CurrentState == State.Aborted) break;

                subCommand.Run();

                ProcessAbortAndPauseEvents();

                // If there is an error or abort in one of the subcommands, terminate the command
                if (subCommand.CurrentState == State.Failed || subCommand.CurrentState == State.Aborted)
                    break;
            }
        }
    }
}
